### PR TITLE
fix(metadata): Add validation tag for UpdateDTO

### PR DESCRIPTION
**Pull request by weichou1229**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/.github/Contributing.md.

## What is the current behavior?

- Update device service with empty name will return status code 404
- It does not allow to update the baseAddress with empty string.

## What is the new behavior?
UpdateDeviceDTO and UpdateDeviceServiceDTO should be able to verify the id whether is valid UUID format and the required name should not be empty.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## New Imports

- [ ] Yes
- [x] No

## Specific Instructions
I tried several tag combinations, like,
- required_without and uuid tag
- required_without and gt=0 tag
- required_without and omitempty and uuid tag
- required_without and omitempty and gt=0 tag

But it didn't pass expected test cases:
- When Id exists, Id should be valid UUID
- When Name exists, Name should not be empty

So I try to write tags  **edgex-update-dto-id**, **edgex-update-dto-id** to verify Id and Name.

Please help review whether the implementation is properly.

### CODE REVIEW

File: v2/validator.go [hunk 1]
```diff
@@ -73,3 +79,29 @@ func ValidateAutoEventFrequency(fl validator.FieldLevel) bool {
...
+	if val.Kind() != reflect.String {
+		return true
+	}
+	// The string value should not be empty
+	if len(strings.TrimSpace(val.String())) > 0 {
+		return true
+	} else {
+		return false
+	}
+}
```
Comment (lenny-intel) [inline v2/validator.go:107]: Since this are `*string` types, should be checking `val` for `nil` first?

And ` val.Kind() != reflect.String` will always be true since the type is `*string`.

File: v2/dtos/deviceservice.go [hunk 2]
```diff
@@ -30,9 +30,9 @@ type DeviceService struct {
 // UpdateDeviceService and its properties are defined in the APIv2 specification:
 // https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/core-metadata/2.x#/UpdateDeviceService
 type UpdateDeviceService struct {
-	Id             *string  `json:"id" validate:"required_without=Name"`
-	Name           *string  `json:"name" validate:"required_without=Id"`
-	BaseAddress    *string  `json:"baseAddress"`
+	Id             *string  `json:"id" validate:"required_without=Name,edgex-update-dto-id"`
+	Name           *string  `json:"name" validate:"required_without=Id,edgex-update-dto-name"`
+	BaseAddress    *string  `json:"baseAddress" validate:"omitempty,gt=0"`
```
Comment (lenny-intel) [inline v2/dtos/deviceservice.go:35]: Why is this needed? what does `gt=0` mean for a `*string` type? Is this way to validate that it isn't an empty string? What about blanks. BaseAddress is a fully qualified URL. Should validate it is a fully qualified URL, it not nil.

File: v2/dtos/device.go [hunk 3]
```diff
@@ -34,15 +34,15 @@ type Device struct {
...
+	Name           *string                       `json:"name" validate:"required_without=Id,edgex-update-dto-name"`
+	Description    *string                       `json:"description" validate:"omitempty,gt=0"`
 	AdminState     *string                       `json:"adminState" validate:"omitempty,oneof='LOCKED' 'UNLOCKED'"`
 	OperatingState *string                       `json:"operatingState" validate:"omitempty,oneof='ENABLED' 'DISABLED'"`
 	LastConnected  *int64                        `json:"lastConnected"`
 	LastReported   *int64                        `json:"lastReported"`
-	ServiceName    *string                       `json:"serviceName"`
-	ProfileName    *string                       `json:"profileName"`
+	ServiceName    *string                       `json:"serviceName" validate:"omitempty,gt=0"`
+	ProfileName    *string                       `json:"profileName" validate:"omitempty,gt=0"`
```
Comment (lenny-intel) [inline v2/dtos/device.go:45]: Why not use `edgex-update-dto-name` instead of `gt=0` so string of blanks is caught.

File: v2/validator.go [hunk 4]
```diff
@@ -73,3 +79,29 @@ func ValidateAutoEventFrequency(fl validator.FieldLevel) bool {
...
+	// Since the Id should be string type, so skip the validation when it isn't a kind of string
+	if val.Kind() != reflect.String {
+		return true
+	}
+	_, err := uuid.Parse(val.String())
+	return err == nil
+}
+
+// ValidateUpdateDtoName used to check the UpdateDTO name pointer value
+func ValidateUpdateDtoName(fl validator.FieldLevel) bool {
```
Comment (lenny-intel) [inline v2/validator.go:95]: I like this and see that it can be expanded to validate the `name` only have valid characters.

**Review by lenny-intel — CHANGES_REQUESTED** [review]

File: v2/validator.go (hunk 4 above)
Comment (cloudxxx8) [reply v2/validator.go:95]: ok, we will add the validation for name when resolving https://github.com/edgexfoundry/go-mod-core-contracts/issues/343

File: v2/validator.go (hunk 1 above)
Comment (weichou1229) [reply v2/validator.go]: > Since this are *string types, should be checking val for nil first?

Yes, I modified the code to check the nil pointer.
```
if val.Kind() == reflect.Ptr && val.IsNil() 
```

> And val.Kind() != reflect.String will always be true since the type is *string

IF the pointer value is nil, then val.Kind() => ptr
IF the pointer value is string, then val.kind() => string

File: v2/dtos/deviceservice.go (hunk 2 above)
Comment (weichou1229) [reply v2/dtos/deviceservice.go]: > Why is this needed? what does gt=0 mean for a *string type? Is this way to validate that it isn't an empty string? What about blanks. 

The gt=0 is used to check string length, but we should check the blank string as you mentioned.

> BaseAddress is a fully qualified URL. Should validate it is a fully qualified URL, it not nil.

Added uri tag for validation. https://pkg.go.dev/github.com/go-playground/validator/v10#hdr-URI_String

File: v2/dtos/device.go (hunk 3 above)
Comment (weichou1229) [reply v2/dtos/device.go]: Modified. You're right, we should check the blank string.

**Review by lenny-intel — APPROVED** [review]
LGTM
